lms/djangoapps/grades/signals/handlers.py

```diff
@@ -244,6 +247,13 @@ def recalculate_course_grade(sender, course, course_structure, user, **kwargs):
     CourseGradeFactory().update(user, course=course, course_structure=course_structure)
 
 
+@receiver(ENROLLMENT_TRACK_UPDATED)
+@receiver(COHORT_MEMBERSHIP_UPDATED)
+def recalculate_course_grade_on_user_partition_change(sender, user, course_key, **kwargs):
+    if waffle().is_enabled(WRITE_ONLY_IF_ENGAGED) or CourseGradeFactory().read(user, course_key=course_key):
```

Comment:
Why are these checks made at this layer?  `CourseGradeFactory.update` already checks for these doesn't it?  It's better to keep these consolidated within the `CourseGradeFactory` module.

Reply:
I added the check here because @iloveagent57 and I decided that even if the switch is off, we still don't want to write empty grades to the database. Without this check, every enrollment will automatically create a course grade for the user. Enrolling along shouldn't create a course grade, we think.

Reply:
[This code](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/grades/new/course_grade_factory.py#L193-L198) in `CourseGradeFactory` already accounts for that case, I believe.

Reply:
In that code, in the case described, `not waffle().is_enabled(WRITE_ONLY_IF_ENGAGED)` will evaluate to `True`, and the grade will get written. We noticed this when a bunch of tests that implicitly assume the absence of a grade in the database failed.